File: backend/scripts/test_relevancy.py
#!/usr/bin/env python3
"""
Test relevance ordering (Python): load opportunities.json and run rank_opportunities_for_user.
Run: python backend/scripts/test_relevancy.py
"""
import json
import sys
from pathlib import Path

_backend = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(_backend))

from matching.user_matching import rank_opportunities_for_user

data_path = _backend / "data" / "opportunities.json"
with open(data_path) as f:
    opportunities = json.load(f)

user_preferences = {
    "causes": ["food_security", "disaster_relief"],
    "regions": ["Africa"],  # broad region -> matched to countries (Kenya, Sudan, etc.) via region_countries.json
    "impact_types": [],
}

ranked = rank_opportunities_for_user(user_preferences, opportunities)

print("Relevancy sort test (Python)")
print("User causes:", user_preferences["causes"])
print("User regions:", user_preferences["regions"])
print("Total opportunities:", len(opportunities))
print("\nTop 10 (relevance-ordered, most recent first within tier):\n")

for i, opp in enumerate(ranked[:10]):
    r = opp.get("region") or ""
    c = opp.get("cause") or ""
    date = (opp.get("date_discovered") or "")[:10]
    title = (opp.get("title") or "")[:50]
    print(f"  {i+1}. {r} | {c} | {date} | {title}...")

print("\n✓ Python ranking: matching region+cause first, then by date.")

File: backend/scripts/test_relevancy.js
/**
 * Test relevance ordering without DB: reads opportunities.json and runs sortByRelevancy.
 * Run: node backend/scripts/test_relevancy.js [causes] [regions]
 * Example: node backend/scripts/test_relevancy.js "food_security,refugees" "Kenya,Global"
 */
const path = require("path");
const fs = require("fs");

const dataPath = path.join(__dirname, "../data/opportunities.json");
const opportunities = JSON.parse(fs.readFileSync(dataPath, "utf8"));

const { sortByRelevancy } = require("../helpers/db/opportunities");

const causesStr = process.argv[2] || "food_security,disaster_relief";
const regionsStr = process.argv[3] || "Kenya,Sudan";
const userCauses = causesStr.split(",").map((s) => s.trim()).filter(Boolean);
const userRegions = regionsStr.split(",").map((s) => s.trim()).filter(Boolean);

const sorted = sortByRelevancy(opportunities, userCauses, userRegions);

console.log("Relevancy sort test (Node)");
console.log("User causes:", userCauses);
console.log("User regions:", userRegions);
console.log("Total opportunities:", opportunities.length);
console.log("\nTop 10 (relevance-ordered, most recent first within tier):\n");

sorted.slice(0, 10).forEach((opp, i) => {
  const r = opp.region || "";
  const c = opp.cause || "";
  const date = (opp.date_discovered || "").slice(0, 10);
  const regionMatch = userRegions.some((ur) => r.toLowerCase().includes(ur.toLowerCase()) || ur.toLowerCase().includes(r.toLowerCase()));
  const causeMatch = userCauses.includes(c) || (opp.values || []).some((v) => userCauses.includes(v));
  const tier = regionMatch && causeMatch ? "region+cause" : regionMatch ? "region" : causeMatch ? "cause" : "—";
  console.log(`  ${i + 1}. [${tier}] ${r} | ${c} | ${date} | ${opp.title?.slice(0, 50)}...`);
});

console.log("\n✓ Relevancy ordering: matching region+cause first, then by date.");
